core/store/src/parallel_iter.rs

```diff
@@ -0,0 +1,692 @@
+/// receiving ends to error, and that's when the worker threads will exit.
+struct WorkQueueItem {
+    work: WorkItem,
+    work_sender: Arc<flume::Sender<WorkQueueItem>>,
+}
+
+impl StoreParallelIterator {
+    /// Creates a new parallel iterator that performs the given task.
+    fn new(options: ParallelIterationOptions, initial_task: WorkItem) -> Self {
+        let (tx, rx) = flume::unbounded::<WorkQueueItem>();
+        let mut threads = Vec::new();
+        let shared_state = Arc::new(ParallelIterationSharedState {
+            num_threads_free: AtomicUsize::new(options.num_threads),
+            thread_work: Mutex::new(vec![None; options.num_threads]),
+        });
+        let tx = Arc::new(tx);
+        tx.send(WorkQueueItem { work: initial_task, work_sender: tx.clone() }).unwrap();
+        drop(tx);
+        for idx in 0..options.num_threads {
+            let rx = rx.clone();
+            let shared_state = shared_state.clone();
+            let options = options.clone();
+            let thread = std::thread::spawn(move || loop {
+                let Ok(WorkQueueItem { work, work_sender }) = rx.recv() else {
+                    // If receiving fails, it means all senders are dropped,
+                    // i.e. there is no more work to do.
+                    break;
+                };
+                shared_state.num_threads_free.fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
+                shared_state.thread_work.lock().unwrap()[idx] = Some(work.clone());
+
+                // Spend some uninterrupted time to work on the task before checking whether we
+                // should split. This is to avoid unnecessary thrashing when the remaining work
+                // is small. Also, randomize this uninterrupted time, so that when there are free
+                // threads, the other threads don't all try to split at the same time.
+                let min_time_to_split = Instant::now()
+                    + Duration::seconds_f64(
+                        options.min_time_before_split.as_seconds_f64()
+                            * (rand::random::<f64>() + 1.0),
+                    );
+
+                match work {
+                    WorkItem::Range(range) => {
+                        let start = if range.start.is_empty() {
+                            None
+                        } else {
+                            Some(range.start.as_slice())
+                        };
+                        let end =
+                            if range.end.is_empty() { None } else { Some(range.end.as_slice()) };
+                        let iter = options.store.iter_range_raw_bytes(options.column, start, end);
+
+                        for kv in iter {
+                            let (key, value) = kv.unwrap();
+                            (options.callback)(&key, Some(&value));
+                            if Instant::now() > min_time_to_split {
```

Comment:
if min_time_to_split is way far in the future, it feels a little sad to me to be putting an Instant::now() and associated syscall in this otherwise tight loop no? im not sure what exactly it is we're trying to avoid/achieve here, but i wonder if theres something better. bc my guess is that calling Instant::now() on every key/val returned by the iterator is actually a pretty non trivial slowdown

Reply:
That's what I thought too, but given how fast the parallel iteration is, it doesn't seem to me it's a noticeable bottleneck.